objc2-foundation-sys/src/generated.rs

```diff
@@ -0,0 +1,245 @@
+//! Example of what I would like `bindgen` to be able to generate.
+
+use core::mem::ManuallyDrop;
+use core::ptr::NonNull;
+
+use crate::NSString;
+use objc2::ffi::NSUInteger;
+use objc2::rc::{Id, Unknown};
+use objc2::runtime::{Bool, Class, Imp, Object, Protocol, Sel};
+use objc2::{class, msg_send, Encoding, Message, RefEncode};
+
+pub type Autoreleased<T> = NonNull<T>;
+pub type NSInvocation = NSObject;
+pub type NSMethodSignature = NSObject;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct _NSZone {
+    _unused: [u8; 0],
+}
+
+unsafe impl RefEncode for _NSZone {
+    const ENCODING_REF: Encoding<'static> = Encoding::Unknown; // TODO
+}
+
+#[repr(transparent)]
+pub struct NSObject(Object);
+
+unsafe impl Message for NSObject {}
+unsafe impl RefEncode for NSObject {
+    const ENCODING_REF: Encoding<'static> = Encoding::Object;
+}
+
+impl std::ops::Deref for NSObject {
+    type Target = Object;
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+impl std::ops::DerefMut for NSObject {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl NSObject {
+    // alloc family
+
+    pub unsafe fn alloc() -> Option<Id<Self, Unknown>> {
+        Id::new_null(msg_send![class!(NSObject), alloc])
```

Comment:
How are you thinking this might work with an inheritance structure? If the `alloc` message doesn't depend on the class, this is fine. My intuition tells me that it depends on the thing in `class!(...)` though.

Reply:
I didn't originally see the reason for it, but what you've done is right (generate an `alloc` method on every class)

Reply:
I realize I probably didn't formulate myself properly here... What I should have said is:

Yup, you're correct, all classes will need an `alloc` method. And that is also what I've done in this "proposal" / idea, `NSString::alloc`, `NSMutableString::alloc`, `NSData::alloc`, ... exists.

Reply:
I added `ClassType::alloc` in #272, so that should be accounted for